### PR TITLE
Enable ODE solvers for longer simulations of trajectories

### PR DESCRIPTION
**Pull request by hvater**
- The possibility of rolling out longer simulations with diffrax ODE solvers is added (in single batch and vmapped form)
- Some clean up of documentation nomenclature and syntax

### CODE REVIEW

**Comment by hvater** [comment]
I would like to have tau as the default step size for sim ahead methods which obviously does not work like this: 

image

Do you have a good idea how we could do this?

**Comment by hvater** [comment]
I think its not very intuitive for users that env_properties needs to be passed to step, but not to vmap_step

image
image

I understand why but this does not really look great.

File: exciting_environments/core_env.py [hunk 1]
```diff
@@ -114,73 +157,192 @@ def create_in_axes_env_properties(self):
...
+        observations = jax.vmap(
+            self.generate_observation, in_axes=(0, self.in_axes_env_properties.physical_constraints)
+        )(states, self.env_properties.physical_constraints)
+
+        # generate rewards - use obs[1:] because obs[0] is observation for the initial state
+        reward = jax.vmap(self.reward_func, in_axes=(0, 0, self.in_axes_env_properties.action_constraints))(
+            observations[1:],
+            jnp.expand_dims(jnp.repeat(actions, int(action_stepsize / obs_stepsize)), 1),
+            self.env_properties.action_constraints,
         )
```
Comment (hvater) [inline exciting_environments/core_env.py:294]: I see no reason to not also compute the reward for the initial observation. If someone does not want to use it, they could still discard it. Wdyt?

File: exciting_environments/cart_pole/cart_pole_env.py [hunk 2]
```diff
@@ -205,26 +208,87 @@ def vector_field(t, y, args):
...
             theta=theta_k1,
             omega=omega_k1,
         )
         opt = None  # Optional(something=...)
-        return self.States(physical_states=phys, PRNGKey=None, optional=opt)
+        return self.State(physical_state=phys, PRNGKey=None, optional=opt)
+
+    @partial(jax.jit, static_argnums=[0, 4, 5])
+    def _ode_solver_simulate_ahead(self, init_state, actions, static_params, obs_stepsize, action_stepsize):
+        """Computes states by simulating a trajectory with given actions."""
```
Comment (OliverSchw) [inline exciting_environments/cart_pole/cart_pole_env.py:222]: probably "state" instead of "states" here as well

File: exciting_environments/core_env.py (hunk 1 above)
Comment (OliverSchw) [reply exciting_environments/core_env.py]: The action is sometimes taken into account in the reward. We wouldn't have any action for the initial observation. Would be a little tricky then. I think we had also discussed that the reward should no longer be calculated at this point anyway and will be calculated later outside the env in the gym_wrapper. I don't know how we want to keep this for the push now.

Comment (hvater) [reply exciting_environments/core_env.py]: Yes I agree. For time being the rewards will not be touched. 
We will have to decide if we want to keep the option for a reward.

As we discussed the rewards should not be part of the internal simulation. Likely they will be an option for the wrapper level.

Comment (hvater) [reply exciting_environments/core_env.py]: Any user that wants to use the framework for RL will likely implement their own reward anyways (?)

**Review by OliverSchw — APPROVED** [review]
